Update DeviceAPI.cs

sscho91 opened the pull request.

jasedit committed:
Cleaning up usings and print help on no arguments.

jasedit committed:
DeviceAPI now protects against bad JSON blobs.

jasedit committed:
Protecting against JSON parse errors in House, plus removing a deprecated function.

Commit:
Update DeviceAPI.cs

Attempt to implement deviceAPI.  Still need a method to pull list from server.

willrd9 committed:
Added IODevices to setters/getters and update

Commit:
Update DeviceAPI.cs

jasedit commented on api/DeviceAPI.cs:
```diff
@@ -34,18 +35,41 @@ public List<string> enumerateDevices(UInt64 house_id)
-	             where device.ID.HouseID == house_id
-	             select device);
-	        return devicelist;
+        if (house_id == 0)
```
ID 0 is a valid House ID.

sscho91 replied:
Ah ok I thought 0 was invalid for some reason
On Apr 23, 2015 10:25 AM, "Jason Ziglar" notifications@github.com wrote:

> In api/DeviceAPI.cs
> https://github.com/ECE4574-5574/Devices/pull/50#discussion_r28966756:
> 
> > @@ -34,18 +35,41 @@ public List<string> enumerateDevices(UInt64 house_id)
> >     {
> >         //TODO: Verify the input parameters are sufficient
> >         //TODO: Implement this function
> > -       if (house_id == 0)
> > -           {
> > -               return null;
> > -           }
> > -           //get device list
> > -           var devicelist = new List<string>();
> > -           var devices = new List<Device>();
> > -           var dlist =
> > -               (from device in devices
> > -                where device.ID.HouseID == house_id
> > -                select device);
> > -           return devicelist;
> > -        if (house_id == 0)
> 
> ID 0 is a valid House ID.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/ECE4574-5574/Devices/pull/50/files#r28966756.

jasedit replied:
Likely because for Room ID's, 0 _is_ invalid. But for the other two parts of the ID, it isn't a magic number.

jasedit commented on api/DeviceAPI.cs:
```diff
@@ -34,18 +35,41 @@ public List<string> enumerateDevices(UInt64 house_id)
+		foreach(var dev in devices)//fake_Devices)
+		{
+			if(house_id == dev.ID.HouseID)
+				devicelist.Add(dev.Name);
```
The string should contain everything needed to communicate with the device (URL, ID, etc,) since this string is passed to the Server in the registration step for storage and future usage.

jasedit committed:
Merge pull request #46 from ECE4574-5574/code_cleanup

Code cleanup

jasedit committed:
Moving DeviceIDOnlyConverter to api for re-use elsewhere.

jasedit committed:
Adding a draft of a formal validation of the House app.

Commit:
Update DeviceAPI.cs

Commit:
Update Test.cs

two tests.  one to show equal desired ID and one to show equal lists but fails because different objects

jasedit committed:
Merge pull request #51 from ECE4574-5574/move_converter

Moving DeviceIDOnlyConverter to api for re-use elsewhere.

jasedit committed:
Expanding DeviceAPI to add client/server specific deserialization calls.

arunkt1 committed:
Merge pull request #53 from ECE4574-5574/add_creation_calls

Expanding DeviceAPI to add client/server specific deserialization calls.

willrd9 committed:
Added bool property for successful update

jasedit committed:
Format hammer.

jasedit committed:
Simplifying getter methods.

jasedit committed:
Adding Device-to-Device updating.

jasedit committed:
Adding default values for DeviceInput/Output.

jasedit committed:
Making Class private, and LastUpdated public. Also making TimeFrame not respond to deviceinput/output.

jasedit committed:
NullDeviceInput now updates time.

jasedit committed:
Fixing unit tests.

jasedit committed:
Merge pull request #49 from ECE4574-5574/PlumbDevices

Added IODevices to setters/getters and update

jasedit committed:
Tweaking unit test to update timestamp a bit more readably.

jasedit committed:
Adding resetIO() function to make altering IO interfaces cleaner.

jasedit commented on api/DeviceAPI.cs:
```diff
@@ -34,18 +36,23 @@ public List<string> enumerateDevices(UInt64 house_id)
-	             where device.ID.HouseID == house_id
-	             select device);
-	        return devicelist;
+        if (house_id < 0)
```
house_id is an unsigned integer value, this check can never return true.

jasedit commented on api/DeviceAPI.cs:
```diff
@@ -60,25 +67,28 @@ public Device registerDevice(string name, UInt64 house_id, string info, UInt64 r
+        }
+        return device;
+
+		
```
This is just breaking formatting.

jasedit commented on api/DeviceAPI.cs:
```diff
@@ -89,16 +99,24 @@ public Device registerDevice(string name, UInt64 house_id, string info, UInt64 r
+
+		if (object.ReferenceEquals(x, null) || object.ReferenceEquals(y, null)) return false;
+
+		return x.ID.HouseID == y.ID.HouseID;
```
This comparison seems incomplete, since devices are not uniquely identified by the HouseID.

jasedit commented on api/DeviceAPI.cs:
```diff
@@ -106,8 +124,36 @@ public bool deleteDevice(Device dev)
 	 */
 	public List<Device> getDevices(UInt64 houseID)
 	{
+		if (houseID < 0)
```
Negative integer can never happen given an unsigned integral type.

jasedit commented on api/DeviceAPI.cs:
```diff
@@ -119,12 +165,20 @@ public List<Device> getDevices(UInt64 houseID)
+		string info = null;
+		foreach(var dev in devices)
+		{
+			if(dev.ID.HouseID == houseID && roomID != 0)
```
roomID can be 0

jasedit commented on api/DeviceAPI.cs:
```diff
@@ -119,12 +165,20 @@ public List<Device> getDevices(UInt64 houseID)
+		{
+			if(dev.ID.HouseID == houseID && roomID != 0)
+				devices.Add(dev);
+			if(roomID == 0)
```
A getDevice() call should never register devices.

jasedit commented on api_tests/Test.cs:
```diff
@@ -226,6 +226,36 @@ public void TestDeviceVars()
+		{
+			device.ID.HouseID = 1;
+			device.ID.RoomID = 2;
+			device.ID.DeviceID = 3;
```
This test is breaking an invariant of the system - all devices in the list are given the exact same ID.

jasedit committed:
Adding DeviceIDOnlyConverter back to build.

jasedit committed:
Merge pull request #59 from ECE4574-5574/fix_dropped_file

Adding DeviceIDOnlyConverter back to build.

jasedit committed:
Fixing missing package reference.

jasedit committed:
Merge pull request #60 from ECE4574-5574/fix_restsharp_package

Fixing missing package reference.

jasedit committed:
Adding PortableRest alongside RestSharp.

jasedit committed:
Merge pull request #61 from ECE4574-5574/add_portablerest

Adding PortableRest alongside RestSharp.

willrd9 committed:
Adding DeviceIO to AlarmSystem.

willrd9 committed:
Added IODevice functionality to CeilingFan.

willrd9 committed:
Added IODevice functionality to GarageDoor.

willrd9 committed:
Added IODevice functionality to LightSwitch.

willrd9 committed:
Added IODevice functionality to Thermostat.

jasedit committed:
Removing PortableRest, which is incompatible with iOS.

jasedit committed:
Merge pull request #63 from ECE4574-5574/IODevicesToAllDevices

Added IODevice functionality

jasedit committed:
Merge pull request #54 from ECE4574-5574/add_house_test

Adding a draft of a formal validation of the House app.

jasedit committed:
Merge pull request #56 from ECE4574-5574/tweak_unit_test

Tweaking unit test to update timestamp a bit more readably.

jasedit committed:
Merge pull request #55 from ECE4574-5574/add_device_update

Adding Device-to-Device updating.

arunkt1 committed:
Merge pull request #64 from ECE4574-5574/disable_restclient

Removing PortableRest, which is incompatible with iOS.

Commit:
Update DeviceAPI.cs

Attempt to create a query list from ServerAPI

jasedit committed:
Merge pull request #58 from ECE4574-5574/reset_device_io

Adding resetIO() function to make altering IO interfaces cleaner.

jasedit committed:
Fixing visibility of UpdateOk property setter.

jasedit committed:
Fixing infinite loops in Device getter/setters.

jasedit committed:
Removing RestSharp.

Commit:
Update DeviceAPI.cs

Need confirmation my approach is ok

jasedit committed:
House app now takes house_id as an integral ID, not a name.
Fixes #67.

jasedit committed:
Adding a simulated thermostat input/output.

jasedit committed:
Adding TimeFrame as argument for creating objects.

jasedit committed:
Merge pull request #65 from ECE4574-5574/fix_device_plumbing

Fix device plumbing

Commit:
Merge pull request #69 from ECE4574-5574/add_time_frame

Adding TimeFrame as argument for creating objects.

jasedit committed:
Merge pull request #68 from ECE4574-5574/house_uses_id

House app now takes house_id as an integral ID, not a name.

jasedit committed:
Adding accessors for input/output, plus making the TimeFrame not ask a remote authority for input.

jasedit committed:
Adding ability to temporarily silence IO during update calls, for internal use.

jasedit committed:
Making the UpdateDevice() call more robust.

jasedit committed:
More typing checks.

jasedit committed:
Making TimeFrame also a command line argument option.

jasedit committed:
Adding device update options.

Commit:
Update DeviceAPI.cs

Need to see if enumerateDevice is correct

Commit:
Update DeviceAPI.cs

Removed extra enumerateDevice

jasedit committed:
Cleaning up so deserializing from the house safely dereferences the device ID.

Note this requires passing in another parameter.

jasedit committed:
Removing RestSharp from api_tests.

jasedit committed:
Adding implicit construction of a Temperature from a double.

jasedit committed:
Minor reformatting of Device.

jasedit committed:
Updating DeviceAPI with more creation/update options.

jasedit committed:
Blocking ID updates in the House.

jasedit committed:
HouseTest now verifies the House API and performance.

jasedit committed:
Adding constructors for Light class.

jasedit committed:
Now testing that read-only properties aren't updated by POST.

jasedit committed:
Removing from_house flag, which is now deduced from type info.

jasedit committed:
Merge pull request #73 from ECE4574-5574/add_sim_thermostat

Add sim thermostat

jasedit commented:
I can't merge this until it's updated to work with the current master.

jasedit committed:
Format hammer.

jasedit commented:
This branch can't be merged, and won't build for me. @sscho91, this needs to be fixed so it builds and passes unit tests before it can be integrated.

jasedit commented:
This PR has become completely broken, and is duplicating many commits in master. This needs to be fixed badly.

jasedit commented:
@sscho91 is this PR deprecated in favor of #79?

jasedit commented:
Wait - I'm getting confused as to which account is which, you may have to ignore my insanity.

sscho91 commented:
I'll look into and discuss it with you later today.  I did not know the big
changes made to deviceAPI

From,
Steven Cho

On Wed, May 6, 2015 at 11:49 AM, Jason Ziglar <notifications@github.com>
wrote:

> Wait - I'm getting confused as to which account is which, you may have to
> ignore my insanity.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ECE4574-5574/Devices/pull/50#issuecomment-99519721>.
>

Commit:
Update DeviceAPI.cs

Commit:
Update DeviceAPI.cs

Changed getDevices(house,id) did not need to register

jasedit commented:
Deprecated in favor of a newer branch.
